reference/test/matrix/csr_kernels.cpp

```diff
@@ -1291,6 +1292,20 @@ TYPED_TEST(Csr, SortUnsortedMatrix)
 }
 
 
+TYPED_TEST(Csr, ExtractsDiagonal)
+{
+    using T = typename TestFixture::value_type;
+    auto matrix = this->mtx3_unsorted->clone();
+    auto diag = matrix->extract_diagonal();
+
+    ASSERT_EQ(diag->get_size()[0], 3);
+    ASSERT_EQ(diag->get_size()[1], 3);
+    ASSERT_EQ(diag->get_values()[0], (T)0.);
```

Comment:
nit: I would prefer if we could avoid C-style casts (same in the other tests)
```suggestion
    ASSERT_EQ(diag->get_values()[0], T{0.});
```